src/Compilers/Core/Portable/Text/ChangedText.cs

```diff
@@ -240,149 +240,171 @@ private static ImmutableArray<TextChangeRange> Merge(IReadOnlyList<ImmutableArra
             return merged;
         }
 
+        /// <summary>
+        /// Merges the new change ranges into the old change ranges, adjusting the new ranges to be with respect to the original text
+        /// (with neither old or new changes applied) instead of with respect to the original text after "old changes" are applied.
+        ///
+        /// This may require splitting, concatenation, etc. of individual change ranges.
+        /// </summary>
         private static ImmutableArray<TextChangeRange> Merge(ImmutableArray<TextChangeRange> oldChanges, ImmutableArray<TextChangeRange> newChanges)
         {
-            var list = new List<TextChangeRange>(oldChanges.Length + newChanges.Length);
+            if (oldChanges.IsEmpty)
+            {
+                return newChanges;
+            }
+            else if (newChanges.IsEmpty)
+            {
+                return oldChanges;
+            }
 
-            int oldIndex = 0;
-            int newIndex = 0;
-            int oldDelta = 0;
+            var builder = ArrayBuilder<TextChangeRange>.GetInstance();
 
-nextNewChange:
-            if (newIndex < newChanges.Length)
-            {
-                var newChange = newChanges[newIndex];
+            var oldChange = oldChanges[0];
+            var newChange = newChanges[0];
 
-nextOldChange:
-                if (oldIndex < oldChanges.Length)
-                {
-                    var oldChange = oldChanges[oldIndex];
+            var oldIndex = 0;
+            var newIndex = 0;
 
-tryAgain:
-                    if (oldChange.Span.Length == 0 && oldChange.NewLength == 0)
-                    {
-                        // old change is a non-change, just ignore it and move on
-                        oldIndex++;
-                        goto nextOldChange;
-                    }
-                    else if (newChange.Span.Length == 0 && newChange.NewLength == 0)
+            // The sum of characters inserted by old changes minus characters deleted by old changes.
+            // This value must be adjusted whenever characters from an old change are added to `builder`.
+            var oldDelta = 0;
+
+            while (true)
+            {
+                if (oldChange.Span.Length == 0 && oldChange.NewLength == 0)
+                {
+                    if (tryGetNext(oldChanges, ref oldIndex, out oldChange)) continue; else break;
+                }
+                else if (newChange.Span.Length == 0 && newChange.NewLength == 0)
+                {
+                    if (tryGetNext(newChanges, ref newIndex, out newChange)) continue; else break;
```

Comment:
>continue; else break [](start = 77, length = 20)

These two code paths are only hit by the `Fuzz()` test in `TextChangeTests`. Should we have specific tests that also hit these code paths as well? The same applies to the `continue` in the next two `else if` branches.

Reply:
Have added the coverage, thanks!